src/network/netdev/wireguard.c

```diff
@@ -0,0 +1,714 @@
+#include "networkd-manager.h"
+#include "wireguard-netlink.h"
+
+static void resolve_endpoints(NetDev *netdev);
+
+static WireguardPeer *wireguard_peer_new(Wireguard *w, unsigned section) {
+        WireguardPeer *peer;
+
+        assert(w);
+
+        if (w->last_peer_section == section && w->peers)
+                return w->peers;
+
+        peer = new0(WireguardPeer, 1);
+        peer->flags = WGPEER_F_REPLACE_ALLOWEDIPS;
+        if (!peer)
+                return NULL;
+
+        LIST_PREPEND(peers, w->peers, peer);
+        w->last_peer_section = section;
+
+        return peer;
+}
+
+static int set_wireguard_interface(NetDev *netdev) {
+        int r;
+        unsigned int i, j;
+        WireguardPeer *peer, *peer_start;
+        WireguardIpmask *mask, *mask_start = NULL;
+        _cleanup_free_ sd_netlink_message *message = NULL;
+        Wireguard *w;
+        uint32_t serial;
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        peer_start = w->peers;
+
+        do {
+          free(message);
+          message = NULL;
+
+          r = sd_genl_message_new(netdev->manager->genl, &message, SD_GENL_WIREGUARD, WG_CMD_SET_DEVICE);
+          if (r < 0)
+                  return log_netdev_error_errno(netdev, r, "Failed to allocate generic netlink message: %m");
+
+          r = sd_netlink_message_append_string(message, WGDEVICE_A_IFNAME, netdev->ifname);
+          if (r < 0)
+                  return log_netdev_error_errno(netdev, r, "Could not append wireguard interface name: %m");
+
+          if (peer_start == w->peers) {
+                  r = sd_netlink_message_append_data(message, WGDEVICE_A_PRIVATE_KEY, &w->private_key, WG_KEY_LEN);
+                  if (r < 0)
+                          return log_netdev_error_errno(netdev, r, "Could not append wireguard private key: %m");
+
+                  r = sd_netlink_message_append_u16(message, WGDEVICE_A_LISTEN_PORT, w->port);
+                  if (r < 0)
+                          return log_netdev_error_errno(netdev, r, "Could not append wireguard port: %m");
+
+                  r = sd_netlink_message_append_u32(message, WGDEVICE_A_FWMARK, w->fwmark);
+                  if (r < 0)
+                          return log_netdev_error_errno(netdev, r, "Could not append wireguard fwmark: %m");
+
+                  r = sd_netlink_message_append_u32(message, WGDEVICE_A_FLAGS, w->flags);
+                  if (r < 0)
+                          return log_netdev_error_errno(netdev, r, "Could not append wireguard flags: %m");
+          }
+
+          r = sd_netlink_message_open_container(message, WGDEVICE_A_PEERS);
+          if (r < 0)
+                  return log_netdev_error_errno(netdev, r, "Could not append wireguard peer attributes: %m");
+
+          i = 0;
+
+          LIST_FOREACH(peers, peer, peer_start) {
+                  r = sd_netlink_message_open_array(message, ++i);
+                  if (r < 0)
+                          break;
```

Comment:
why the different error handling here? i mean, this function logs about all errors apparently, so it should also log about this one here, no? (and below the same)

Reply:
The errors are supposed to happen. The structure of this function is that it tries to stick as much in a message as it can. When it fails, it breaks, sends what it could fit in, and then constructs the next message containing the next set of data. So it's not an error that's an error -- it's an error code indicating that it's time to send data off and begin forming the next message.

Reply:
yes, this is needed for message chunking.